{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "root_dir = \"../../Downloads/GENOME540/\"\n",
    "state1_counts_path = root_dir + \"STATE1_anc_rep_counts.txt\"\n",
    "state2_counts_path = root_dir + \"STATE2_codon1_2_counts.txt\"\n",
    "align_path = root_dir + \"ENm010.txt\"\n",
    "test_align_path = root_dir + \"ENm006_short.txt\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_counts(filename):\n",
    "    with open(filename, 'r') as f:\n",
    "        data = f.readlines()\n",
    "        data = [x.strip().split(\"\\t\") for x in data]\n",
    "        data = [[x[0], int(x[1])] for x in data]\n",
    "        counts = {}\n",
    "        for x in data:\n",
    "            counts[x[0]] = x[1]\n",
    "        return counts\n",
    "\n",
    "state1_counts = read_counts(state1_counts_path)\n",
    "state2_counts = read_counts(state2_counts_path)\n",
    "\n",
    "symbol_to_int = {}\n",
    "for i, symbol in enumerate(state1_counts.keys()):\n",
    "    symbol_to_int[symbol] = i"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_frequencies(counts):\n",
    "    return {x: counts[x] / sum(counts.values()) for x in counts}\n",
    "\n",
    "state1_freqs = get_frequencies(state1_counts)\n",
    "state2_freqs = get_frequencies(state2_counts)\n",
    "del state1_counts, state2_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "\n",
    "a = np.array([[0.95, 0.05], [0.10, 0.90]]) # transition probabilities\n",
    "b = np.array([list(state1_freqs.values()), list(state2_freqs.values())]) # emission probabilities\n",
    "pi = np.array([0.95, 0.05]) # initiation probabilities\n",
    "del state1_freqs, state2_freqs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Initial State Probabilities:\n",
      "1=0.95000\n",
      "2=0.05000\n",
      "\n",
      "Transition Probabilities:\n",
      "1,1=0.95000\n",
      "1,2=0.05000\n",
      "2,1=0.10000\n",
      "2,2=0.90000\n",
      "\n",
      "Emission Probabilities:\n",
      "1,A--=0.15403\n",
      "1,A-A=0.00725\n",
      "1,A-C=0.00633\n",
      "1,A-G=0.02133\n",
      "1,A-T=0.00412\n",
      "1,AA-=0.01285\n",
      "1,AAA=0.00270\n",
      "1,AAC=0.00150\n",
      "1,AAG=0.00253\n",
      "1,AAT=0.00045\n",
      "1,AC-=0.01318\n",
      "1,ACA=0.00171\n",
      "1,ACC=0.00333\n",
      "1,ACG=0.00280\n",
      "1,ACT=0.00049\n",
      "1,AG-=0.03189\n",
      "1,AGA=0.00211\n",
      "1,AGC=0.00198\n",
      "1,AGG=0.01328\n",
      "1,AGT=0.00111\n",
      "1,AT-=0.01146\n",
      "1,ATA=0.00086\n",
      "1,ATC=0.00085\n",
      "1,ATG=0.00235\n",
      "1,ATT=0.00059\n",
      "1,C--=0.00335\n",
      "1,C-A=0.00864\n",
      "1,C-C=0.00263\n",
      "1,C-G=0.00132\n",
      "1,C-T=0.00036\n",
      "1,CA-=0.00417\n",
      "1,CAA=0.08468\n",
      "1,CAC=0.01749\n",
      "1,CAG=0.00379\n",
      "1,CAT=0.00233\n",
      "1,CC-=0.00203\n",
      "1,CCA=0.02829\n",
      "1,CCC=0.01342\n",
      "1,CCG=0.00181\n",
      "1,CCT=0.00101\n",
      "1,CG-=0.00127\n",
      "1,CGA=0.00689\n",
      "1,CGC=0.00216\n",
      "1,CGG=0.00203\n",
      "1,CGT=0.00028\n",
      "1,CT-=0.00072\n",
      "1,CTA=0.00699\n",
      "1,CTC=0.00210\n",
      "1,CTG=0.00058\n",
      "1,CTT=0.00036\n",
      "1,G--=0.00332\n",
      "1,G-A=0.00282\n",
      "1,G-C=0.01320\n",
      "1,G-G=0.00171\n",
      "1,G-T=0.00049\n",
      "1,GA-=0.00198\n",
      "1,GAA=0.01328\n",
      "1,GAC=0.03190\n",
      "1,GAG=0.00211\n",
      "1,GAT=0.00111\n",
      "1,GC-=0.00635\n",
      "1,GCA=0.02137\n",
      "1,GCC=0.15465\n",
      "1,GCG=0.00726\n",
      "1,GCT=0.00409\n",
      "1,GG-=0.00150\n",
      "1,GGA=0.00252\n",
      "1,GGC=0.01285\n",
      "1,GGG=0.00271\n",
      "1,GGT=0.00045\n",
      "1,GT-=0.00084\n",
      "1,GTA=0.00235\n",
      "1,GTC=0.01149\n",
      "1,GTG=0.00086\n",
      "1,GTT=0.00057\n",
      "1,T--=0.01338\n",
      "1,T-A=0.00181\n",
      "1,T-C=0.00203\n",
      "1,T-G=0.02824\n",
      "1,T-T=0.00099\n",
      "1,TA-=0.00216\n",
      "1,TAA=0.00202\n",
      "1,TAC=0.00127\n",
      "1,TAG=0.00687\n",
      "1,TAT=0.00028\n",
      "1,TC-=0.00262\n",
      "1,TCA=0.00132\n",
      "1,TCC=0.00335\n",
      "1,TCG=0.00865\n",
      "1,TCT=0.00036\n",
      "1,TG-=0.01747\n",
      "1,TGA=0.00380\n",
      "1,TGC=0.00417\n",
      "1,TGG=0.08460\n",
      "1,TGT=0.00231\n",
      "1,TT-=0.00211\n",
      "1,TTA=0.00059\n",
      "1,TTC=0.00072\n",
      "1,TTG=0.00698\n",
      "1,TTT=0.00035\n",
      "2,A--=0.22082\n",
      "2,A-A=0.00198\n",
      "2,A-C=0.00101\n",
      "2,A-G=0.00524\n",
      "2,A-T=0.00030\n",
      "2,AA-=0.00309\n",
      "2,AAA=0.00113\n",
      "2,AAC=0.00021\n",
      "2,AAG=0.00048\n",
      "2,AAT=0.00003\n",
      "2,AC-=0.00203\n",
      "2,ACA=0.00027\n",
      "2,ACC=0.00069\n",
      "2,ACG=0.00041\n",
      "2,ACT=0.00003\n",
      "2,AG-=0.00761\n",
      "2,AGA=0.00041\n",
      "2,AGC=0.00025\n",
      "2,AGG=0.00469\n",
      "2,AGT=0.00007\n",
      "2,AT-=0.00058\n",
      "2,ATA=0.00005\n",
      "2,ATC=0.00003\n",
      "2,ATG=0.00013\n",
      "2,ATT=0.00012\n",
      "2,C--=0.00112\n",
      "2,C-A=0.00316\n",
      "2,C-C=0.00035\n",
      "2,C-G=0.00036\n",
      "2,C-T=0.00003\n",
      "2,CA-=0.00142\n",
      "2,CAA=0.21487\n",
      "2,CAC=0.00527\n",
      "2,CAG=0.00191\n",
      "2,CAT=0.00039\n",
      "2,CC-=0.00028\n",
      "2,CCA=0.00917\n",
      "2,CCC=0.00352\n",
      "2,CCG=0.00043\n",
      "2,CCT=0.00009\n",
      "2,CG-=0.00028\n",
      "2,CGA=0.00301\n",
      "2,CGC=0.00042\n",
      "2,CGG=0.00123\n",
      "2,CGT=0.00004\n",
      "2,CT-=0.00004\n",
      "2,CTA=0.00076\n",
      "2,CTC=0.00011\n",
      "2,CTG=0.00006\n",
      "2,CTT=0.00016\n",
      "2,G--=0.00069\n",
      "2,G-A=0.00040\n",
      "2,G-C=0.00202\n",
      "2,G-G=0.00026\n",
      "2,G-T=0.00003\n",
      "2,GA-=0.00024\n",
      "2,GAA=0.00475\n",
      "2,GAC=0.00764\n",
      "2,GAG=0.00041\n",
      "2,GAT=0.00007\n",
      "2,GC-=0.00102\n",
      "2,GCA=0.00520\n",
      "2,GCC=0.22090\n",
      "2,GCG=0.00199\n",
      "2,GCT=0.00031\n",
      "2,GG-=0.00022\n",
      "2,GGA=0.00047\n",
      "2,GGC=0.00304\n",
      "2,GGG=0.00113\n",
      "2,GGT=0.00003\n",
      "2,GT-=0.00003\n",
      "2,GTA=0.00014\n",
      "2,GTC=0.00059\n",
      "2,GTG=0.00005\n",
      "2,GTT=0.00013\n",
      "2,T--=0.00351\n",
      "2,T-A=0.00043\n",
      "2,T-C=0.00029\n",
      "2,T-G=0.00914\n",
      "2,T-T=0.00009\n",
      "2,TA-=0.00041\n",
      "2,TAA=0.00123\n",
      "2,TAC=0.00028\n",
      "2,TAG=0.00301\n",
      "2,TAT=0.00004\n",
      "2,TC-=0.00035\n",
      "2,TCA=0.00034\n",
      "2,TCC=0.00110\n",
      "2,TCG=0.00315\n",
      "2,TCT=0.00003\n",
      "2,TG-=0.00525\n",
      "2,TGA=0.00190\n",
      "2,TGC=0.00142\n",
      "2,TGG=0.21461\n",
      "2,TGT=0.00038\n",
      "2,TT-=0.00011\n",
      "2,TTA=0.00006\n",
      "2,TTC=0.00004\n",
      "2,TTG=0.00076\n",
      "2,TTT=0.00016\n"
     ]
    }
   ],
   "source": [
    "print(\"Initial State Probabilities:\")\n",
    "print(\"1={:.5f}\".format(pi[0]))\n",
    "print(\"2={:.5f}\".format(pi[1]))\n",
    "\n",
    "print(\"\\nTransition Probabilities:\")\n",
    "print(\"1,1={:.5f}\".format(a[0,0]))\n",
    "print(\"1,2={:.5f}\".format(a[0,1]))\n",
    "print(\"2,1={:.5f}\".format(a[1,0]))\n",
    "print(\"2,2={:.5f}\".format(a[1,1]))\n",
    "\n",
    "print(\"\\nEmission Probabilities:\")\n",
    "for i, symbol in enumerate(sorted(symbol_to_int.keys())):\n",
    "    print(\"1,{}={:.5f}\".format(symbol, b[0, i]))\n",
    "for i, symbol in enumerate(sorted(symbol_to_int.keys())):\n",
    "    print(\"2,{}={:.5f}\".format(symbol, b[1, i]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_alignment(filename):\n",
    "    with open(filename, 'r') as f:\n",
    "        return f.read().split(\"\\n\")\n",
    "\n",
    "def parse_alignment(filename):\n",
    "    lines = read_alignment(filename)\n",
    "    start_pos = int(lines[0].split('-')[0].split(':')[1])\n",
    "    sequences = {}\n",
    "    for line in lines:\n",
    "        if line and not line.startswith('#'):\n",
    "            species, sequence = line.split('\\t')\n",
    "            if species not in sequences:\n",
    "                sequences[species] = ''\n",
    "            sequences[species] += sequence\n",
    "    sequences = [sequences[species] for species in sequences]\n",
    "    sequences = [''.join(column) for column in zip(*sequences)]\n",
    "    sequences = [symbol_to_int[s] for s in sequences]\n",
    "    return start_pos, np.array(sequences)\n",
    "\n",
    "START, ALIGNMENTS = parse_alignment(align_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "class HMM(object):\n",
    "\n",
    "    def __init__(self, a, b, pi):\n",
    "        self.A = a\n",
    "        self.B = b\n",
    "        self.pi = pi\n",
    "    \n",
    "    def viterbi(self, V):\n",
    "\n",
    "        T = V.shape[0]\n",
    "        M = self.A.shape[0]\n",
    "\n",
    "        # delta and psi\n",
    "        delta = np.zeros((T, M))\n",
    "        psi = np.zeros((T, M))\n",
    "\n",
    "        # delta for t=0\n",
    "        delta[0, :] = np.log(self.pi) + np.log(self.B[:, V[0]])\n",
    "\n",
    "        # delta and psi for t > 0\n",
    "        for t in range(1, T):\n",
    "            for j in range(M):\n",
    "                probs = delta[t - 1] + np.log(self.A[:, j]) + np.log(self.B[j, V[t]])\n",
    "                delta[t, j] = np.max(probs)\n",
    "                psi[t, j] = np.argmax(probs)\n",
    "\n",
    "        # backtrack to find optimal path\n",
    "        path = np.zeros(T, dtype=int)\n",
    "        path[T-1] = np.argmax(delta[T-1, :])\n",
    "        for t in range(T-2, -1, -1):\n",
    "            path[t] = psi[t+1, path[t+1]]\n",
    "\n",
    "        return path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "hmm = HMM(a, b, pi)\n",
    "path = hmm.viterbi(ALIGNMENTS)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "State Histogram:\n",
      "1=221617\n",
      "2=54130\n",
      "\n",
      "Segment Histogram:\n",
      "1=912\n",
      "2=911\n"
     ]
    }
   ],
   "source": [
    "def count_segs(arr):\n",
    "    diffs = np.diff(arr)\n",
    "    n_ones = np.sum(diffs == 1) + (arr[0] == 1)\n",
    "    n_zeros = np.sum(diffs == -1) + (arr[0] == 0)\n",
    "    return n_zeros, n_ones\n",
    "\n",
    "print(\"State Histogram:\")\n",
    "print(\"1=\"+str(np.sum(path == 0)))\n",
    "print(\"2=\"+str(np.sum(path == 1)))\n",
    "\n",
    "print(\"\\nSegment Histogram:\")\n",
    "print(\"1=\"+str(count_segs(path)[0]))\n",
    "print(\"2=\"+str(count_segs(path)[1]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Longest Segment List:\n",
      "\n",
      "27106783 27108023\n",
      "27149114 27150259\n",
      "27170930 27171941\n",
      "27116222 27117125\n",
      "27108299 27109088\n",
      "27190694 27191482\n",
      "27189995 27190665\n",
      "27153545 27154205\n",
      "27180046 27180696\n",
      "26934893 26935504\n"
     ]
    }
   ],
   "source": [
    "def find_longest_sequences(arr, n):\n",
    "    \n",
    "    sequences = []\n",
    "    current_sequence = 0\n",
    "\n",
    "    for i in range(len(arr)):\n",
    "        if arr[i] == 1:\n",
    "            current_sequence += 1\n",
    "        else:\n",
    "            if current_sequence > 0:\n",
    "                sequences.append((i - current_sequence, current_sequence-1))\n",
    "                current_sequence = 0\n",
    "                \n",
    "    if current_sequence > 0:\n",
    "        sequences.append((len(arr) - current_sequence, current_sequence-1))\n",
    "\n",
    "    sequences.sort(key=lambda x: x[1], reverse=True)\n",
    "    sequences = [(x[0]+START, x[0]+START+x[1]) for x in sequences]\n",
    "\n",
    "    return sequences[:n]\n",
    "\n",
    "top10 = find_longest_sequences(path, 10)\n",
    "print(\"Longest Segment List:\\n\")\n",
    "for start, end in top10:\n",
    "    print(start, end)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Annotations:\n",
      "\n",
      "Start: 27106783\n",
      "End: 27108023 \n",
      "\n",
      "\n",
      "Start: 27149114\n",
      "End: 27150259 \n",
      "\n",
      "\n",
      "Start: 27170930\n",
      "End: 27171941 \n",
      "\n",
      "\n",
      "Start: 27116222\n",
      "End: 27117125 \n",
      "\n",
      "\n",
      "Start: 27108299\n",
      "End: 27109088 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(\"Annotations:\")\n",
    "\n",
    "for seq in top10[:5]:\n",
    "    print(\"\\nStart:\", seq[0])\n",
    "    print(\"End:\", seq[1], \"\\n\")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Descriptions from UCSC genome browser --> https://genome.ucsc.edu/cgi-bin/hgGateway\n",
    "\n",
    "Position/Search Term\n",
    "\n",
    "chr7: 27106783-27108023 --> HOXA2 gene --> This gene codes for a protein called homeobox A2, which is a transcription factor that plays a role in embryonic development, particularly in the formation of the hindbrain and facial structures.\n",
    "\n",
    "chr7: 27149114-27150259 --> HOXA5 gene --> This gene codes for a protein called homeobox A5, which is also a transcription factor involved in embryonic development. HOXA5 is important for the development of various organs and tissues, including the limbs, respiratory system, and urogenital system.\n",
    "\n",
    "chr7: 27170930-27171941 --> HOXA9 gene --> This gene codes for a protein called homeobox A9, which is another transcription factor involved in embryonic development. HOXA9 is particularly important for the development of blood cells and is frequently dysregulated in leukemia.\n",
    "\n",
    "chr7: 27116222-27117125 --> HOXA3 gene --> This gene codes for a protein called homeobox A3, which is a transcription factor that plays a role in embryonic development, particularly in the development of the central nervous system and the gut.\n",
    "\n",
    "chr7: 27108299-27109088 --> HOXA2 gene again"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "deep-learning-env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.15"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "31c816c56b3f001d803f5818cb31842c102a89c21941618a472b08849d47f2d3"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
